test/matrix/matrix.cpp

```diff
@@ -0,0 +1,782 @@
+    using mixed_value_type = gko::next_precision<value_type>;
+    using Vec = gko::matrix::Dense<value_type>;
+    using MixedVec = gko::matrix::Dense<mixed_value_type>;
+
+    Matrix() : rand_engine(15) {}
+
+    void SetUp()
+    {
+        ref = gko::ReferenceExecutor::create();
+        init_executor(ref, exec);
+    }
+
+    void TearDown()
+    {
+        if (exec != nullptr) {
+            ASSERT_NO_THROW(exec->synchronize());
+        }
+    }
+
+    template <typename DistType>
+    gko::matrix_data<value_type, index_type> gen_mtx_data(int num_rows,
+                                                          int num_cols,
+                                                          DistType dist)
+    {
+        return gko::test::generate_random_matrix_data<value_type, index_type>(
+            num_rows, num_cols, dist, std::normal_distribution<>(0.0, 1.0),
+            rand_engine);
+    }
+
+    gko::matrix_data<value_type, index_type> gen_mtx_data(int num_rows,
+                                                          int num_cols,
+                                                          int min_cols,
+                                                          int max_cols)
+    {
+        return gen_mtx_data(
+            num_rows, num_cols,
+            std::uniform_int_distribution<>(min_cols, max_cols));
+    }
+
+    template <typename ValueType, typename IndexType>
+    gko::matrix_data<ValueType, IndexType> gen_dense_data(gko::dim<2> size)
+    {
+        return {size, std::normal_distribution<>(0.0, 1.0), rand_engine};
+    }
+
+    template <typename VecType = Vec>
+    test_pair<VecType> gen_in_vec(const test_pair<Mtx>& mtx, int nrhs,
+                                  int stride)
+    {
+        auto size = gko::dim<2>{mtx.ref->get_size()[1],
+                                static_cast<gko::size_type>(nrhs)};
+        auto result = VecType::create(ref, size, stride);
+        result->read(gen_dense_data<typename VecType::value_type,
+                                    typename Mtx::index_type>(size));
+        return {std::move(result), exec};
+    }
+
+    template <typename VecType = Vec>
+    test_pair<VecType> gen_scalar()
+    {
+        return {gko::initialize<VecType>(
+                    {gko::test::detail::get_rand_value<
+                        typename VecType::value_type>(
+                        std::normal_distribution<>(0.0, 1.0), rand_engine)},
+                    ref),
+                exec};
+    }
+
+    template <typename VecType = Vec>
+    test_pair<VecType> gen_out_vec(const test_pair<Mtx>& mtx, int nrhs,
+                                   int stride)
+    {
+        auto size = gko::dim<2>{mtx.ref->get_size()[0],
+                                static_cast<gko::size_type>(nrhs)};
+        auto result = VecType::create(ref, size, stride);
+        result->read(gen_dense_data<typename VecType::value_type,
+                                    typename Mtx::index_type>(size));
```

Comment:
how about to add the NaN in output vector when they are used in apply(b, x)?
It may give the issue on dpcpp sparsely csr apply which I do not know they fix or not, or some other formats. But it is easy to fix on our side.